### PR TITLE
#4993: Unpad Height Output Sharded

### PR DESCRIPTION
**Pull request by ntarafdar**
Supports Height Output Sharding

### CODE REVIEW

File: tt_eager/tt_dnn/op_library/work_split.hpp [hunk 1]
```diff
@@ -189,12 +204,14 @@ inline std::tuple<uint32_t, CoreRangeSet, CoreRangeSet, CoreRangeSet, uint32_t,
...
 	uint32_t num_cores_x = grid_size.x, num_cores_y = grid_size.y;
 	auto target_num_cores = std::min(units_to_divide, num_cores_x * num_cores_y);
-	CoreRangeSet all_cores(num_cores_to_corerange_set(target_num_cores, grid_size, row_wise));
 
 	std::set<CoreRange> core_group_1_set;
 	std::set<CoreRange> core_group_2_set;
 	uint32_t units_per_core_group_1 = units_to_divide / target_num_cores;
 	uint32_t units_per_core_group_2 = 0;
+
+    CoreRangeSet all_cores(num_cores_to_corerange_set(target_num_cores, grid_size, row_wise));
```
Comment (TT-billteng) [inline tt_eager/tt_dnn/op_library/work_split.hpp:213]: seems like indenting is a little off

File: tt_metal/impl/buffers/buffer.cpp [hunk 2]
```diff
@@ -157,7 +157,7 @@ Buffer::Buffer(Device *device, uint64_t size, uint64_t page_size, const BufferTy
     validate_buffer_size_and_page_size(size, page_size, buffer_type, buffer_layout, shard_parameters);
     if(is_sharded(buffer_layout)){
```
Comment (TT-billteng) [inline tt_metal/impl/buffers/buffer.cpp:158]: Can't remember if we talked about this, but are you planning to create a separate class for Sharded buffers?

**Review by tt-aho — APPROVED** [review]

File: tt_metal/impl/dispatch/command_queue.cpp [hunk 3]
```diff
@@ -543,8 +543,15 @@ const DeviceCommand EnqueueReadShardedBufferCommand::create_buffer_transfer_inst
 
     uint32_t num_cores = this->buffer.num_cores();
     uint32_t shard_size = this->buffer.shard_spec().size();
-    //TODO: for now all shards are same size of pages
-    vector<uint32_t> num_pages_in_shards(num_cores, shard_size);
+    //TODO: for now all shards are same size of pages except last core
+    vector<uint32_t> num_pages_in_shards(num_cores - 1, shard_size);
+    if(this->buffer.num_pages() == num_cores*shard_size) {
```
Comment (DrJessop) [inline tt_metal/impl/dispatch/command_queue.cpp:548]: spaces between condition and parentheses, variables and *

File: tt_metal/impl/dispatch/command_queue.cpp [hunk 4]
```diff
@@ -543,8 +543,15 @@ const DeviceCommand EnqueueReadShardedBufferCommand::create_buffer_transfer_inst
...
+        num_pages_in_shards.push_back(shard_size);
+    }
+    else {
+        num_pages_in_shards.push_back(this->buffer.num_pages()%shard_size);
```
Comment (DrJessop) [inline tt_metal/impl/dispatch/command_queue.cpp:552]: space between variables and operator %

File: tt_metal/impl/dispatch/command_queue.cpp [hunk 5]
```diff
@@ -688,8 +695,15 @@ const DeviceCommand EnqueueWriteShardedBufferCommand::create_buffer_transfer_ins
...
-    //TODO: for now all shards are same size of pages
-    vector<uint32_t> num_pages_in_shards(num_cores, shard_size);
+    //TODO: for now all shards are same size of pages, except last core
+    vector<uint32_t> num_pages_in_shards(num_cores -1, shard_size);
+    if(this->buffer.num_pages() == num_cores*shard_size) {
+        num_pages_in_shards.push_back(shard_size);
+    }
+    else {
+        num_pages_in_shards.push_back(this->buffer.num_pages()%shard_size);
+    }
```
Comment (DrJessop) [inline tt_metal/impl/dispatch/command_queue.cpp:705]: spaces

File: tt_metal/impl/dispatch/command_queue.cpp [hunk 6]
```diff
@@ -688,8 +695,15 @@ const DeviceCommand EnqueueWriteShardedBufferCommand::create_buffer_transfer_ins
 
     uint32_t num_cores = this->buffer.num_cores();
     uint32_t shard_size = this->buffer.shard_spec().size();
-    //TODO: for now all shards are same size of pages
-    vector<uint32_t> num_pages_in_shards(num_cores, shard_size);
+    //TODO: for now all shards are same size of pages, except last core
+    vector<uint32_t> num_pages_in_shards(num_cores -1, shard_size);
```
Comment (DrJessop) [inline tt_metal/impl/dispatch/command_queue.cpp:699]: num_cores - 1 instead of num_cores -1

Comment (DrJessop) [reply tt_metal/impl/dispatch/command_queue.cpp:699]: we are tying to get rid of push_back code and reserve the amount of space you need AOT. Cleaner to do 

```c++
vector<uint32_t> num_pages_in_shards(num_cores, shard_size);
if (edge_case_logic) {
    num_pages_in_shards[num_pages_in_shards.size() - 1] = edge_case_value;
}
```

Also, for readability, shard_size units are ambiguous. Typically I think of size in bytes in dispatch land, but it seems here that it means num pages in a shard (maybe make that more explicit).

Also, modulo is expensive. If you know num_pages() <= num_cores * shard_size, a cheaper way to compute the edge case is 

```c++
this->buffer.num_pages() - (num_cores - 1) * shard_size;
```

So all in all, the code would look like this

```c++
vector<uint32_t> num_pages_in_shards(num_cores, shard_size);
uint32_t num_pages_in_shards_upper_bound = num_cores * shard_num_pages;
if (this->buffer.num_pages() != num_pages_in_shards_upper_bound) {
   num_pages_in_shards[num_pages_in_shards.size() - 1] = this->buffer.num_pages() - (num_pages_in_shard_upper_bound - shard_size);
}
```

**Review by DrJessop — CHANGES_REQUESTED** [review]

**Review by DrJessop — APPROVED** [review]

**Review by farbabi — APPROVED** [review]

File: tt_metal/impl/buffers/buffer.cpp (hunk 2 above)
Comment (tarafdarTT) [reply tt_metal/impl/buffers/buffer.cpp:158]: ya will do in another PR. I'll file an issue

File: tt_eager/tensor/tensor.cpp [hunk 7]
```diff
@@ -337,9 +337,11 @@ Tensor create_sharded_device_tensor(const Shape& shape, DataType data_type, Layo
 
     ShardSpecBuffer shard_spec_buffer(shard_spec, page_shape, tensor2d_size);
     uint32_t shard_size = shard_shape[0] * shard_shape[1] * element_size;
-    if(layout == Layout::TILE)
+    if(layout == Layout::TILE) {
         shard_size = tensor_impl::packed_buffer_size_bytes_wrapper(data_type, compute_buffer_size(Shape({shard_shape[0], shard_shape[1]}), data_type));
-    uint32_t packed_size_in_bytes = shard_size * num_cores;
+    }
+    //uint32_t packed_size_in_bytes = shard_size * num_cores;
```
Comment (cfjchu) [inline tt_eager/tensor/tensor.cpp:343]: delete?

File: tt_eager/tensor/tensor.cpp [hunk 8]
```diff
@@ -337,9 +337,11 @@ Tensor create_sharded_device_tensor(const Shape& shape, DataType data_type, Layo
 
     ShardSpecBuffer shard_spec_buffer(shard_spec, page_shape, tensor2d_size);
     uint32_t shard_size = shard_shape[0] * shard_shape[1] * element_size;
-    if(layout == Layout::TILE)
+    if(layout == Layout::TILE) {
         shard_size = tensor_impl::packed_buffer_size_bytes_wrapper(data_type, compute_buffer_size(Shape({shard_shape[0], shard_shape[1]}), data_type));
```
Comment (cfjchu) [inline tt_eager/tensor/tensor.cpp:341]: I don't see this `shard_size` variable used at all anymore in this function? Do we need any of the code related to this?

File: tt_eager/tensor/tensor.cpp [hunk 9]
```diff
@@ -231,7 +231,7 @@ bool Tensor::is_allocated() const {
 
 std::vector<uint32_t> Tensor::host_page_ordering(){
     auto cores = buffer()->all_cores();
-    auto shard_size = buffer()->shard_spec().size();
+    auto shard_size = buffer()->shard_spec().num_pages();
```
Comment (cfjchu) [inline tt_eager/tensor/tensor.cpp:234]: is there a better variable name for this now?

**Review by davorchap — APPROVED** [review]

**Review by boris-drazic — APPROVED** [review]

**Review by arakhmati — APPROVED** [review]

**Review by arakhmati — APPROVED** [review]

**Pull request closed without merging**